### PR TITLE
Move lxd profile watcher from model cache to apiserver layer

### PR DESCRIPTION
**Pull request by wallyworld**
TL;DR: the model cache is no longer used for the lxd profile watcher

The implications are that the uniter facade no longer waits for the cache to be updated when setting the charm url, and the lxd profile watcher is driven off state watchers, not the cache.
Also, the charm config settings bypass the model cache, which means generations (which is behind a flag) is broken for now.

The vibe of the change can be seen by the adjustments to the instance mutator facade to plug in the state backed implementation as opposed to the model cache.

The lxd profile watcher is moved from the model cache to the instance mutator facade. The `lxdprofilewather.go` implementation is moved essentially as is, and the core logic is retained. The event feeds for application charm URL change, unit add/remove, machine provisioned, are all retained, but these are now state watchers instead of pubsub topics. The callback methods which used to be plugged into pubsub subscribe are retained and tweaked to handle interfaces for Unit, Application etc backed by state entities.

The existing cache of known app / unit info is retained in the profile watcher. This allows the state watcher to (for now) be too verbose in the any application change is processed; it's a simple tweak to enhance the watcher to only trigger on actual charm URL changes. But because the updated app is compared with what's held in the cache, this can be deferred for now.

In addition, the metrics which had been recorded in the mode cache about the operation of the profile watcher have been removed. These can be added later if needed.

## QA

```
$ ./main.sh -l test deploy test_deploy_charms

==> Checking for dependencies
==> Running subtest test_deploy_charms for deploy suite
==> TEST BEGIN: deploy (tmp.MsI)
==> Checking for dependencies
====> Reusing bootstrapped juju (2.9.25:aws)
====> Bootstrapped juju (2s)
===> [ ✔ ] Success: deploy charm (186s)
===> [ ✔ ] Success: deploy specific series (25s)
===> [ ✔ ] Success: deploy lxd to container (344s)
===> [ ✔ ] Success: deploy lxd profile charm container (339s)
==> TEST SKIPPED: deploy_lxd_to_machine - tests for LXD only
==> TEST SKIPPED: deploy_lxd_profile_charm - tests for LXD only
==> TEST SKIPPED: deploy_local_lxd_profile_charm - tests for LXD only
==> TEST SKIPPED: deploy bundles
==> TEST SKIPPED: CMR bundle deploy tests
==> TEST SKIPPED: deploy to os
==> TEST SKIPPED: deploy revision
====> Destroying model (test-deploy-ctl)
====> Destroyed model (test-deploy-ctl)
==> TEST DONE: deploy (904s)
==> Cleaning up
====> Cleaning up jujus
====> Completed cleaning up jujus

==> Test result: success
==> Tests Removed: /home/ian/juju/go/src/juju/juju/tests/tmp.MsI
==> TEST COMPLETE

```

### CODE REVIEW

File: apiserver/facades/agent/instancemutater/instancemutater.go [hunk 1]
```diff
@@ -51,13 +47,22 @@ type InstanceMutaterAPI struct {
...
 	machineFunc EntityMachineFunc
 }
 
+type InstanceMutatorWatcher interface {
+	WatchLXDProfileVerificationForMachine(id string) (*MachineLXDProfileWatcher, error)
+}
+
+type instanceMutatorWatcher struct {
+	st      InstanceMutaterState
+	metrics *cache.ControllerGauges
```
Comment (manadart) [inline apiserver/facades/agent/instancemutater/instancemutater.go:63]: I say remove all the metrics, including `LXDProfileChangeError`, `LXDProfileChangeNotification` and `LXDProfileNoChange` from _core/cache/metrics.go_ and look at adding them later.

File: apiserver/facades/agent/instancemutater/lxdprofilewatcher_test.go [hunk 2]
```diff
@@ -0,0 +1,476 @@
+// Copyright 2019 Canonical Ltd.
```
Comment (wallyworld) [inline apiserver/facades/agent/instancemutater/lxdprofilewatcher_test.go:1]: moved from `core/cache/lxdprofilewatcher_test.go`
core tests are the same but ported to use the new go mocks

**Review by manadart — APPROVED** [review]
Looks good.

I tested with testcharms/charm-repo/quantal/lxd-profile.
- Worked to apply the profile.
- Upgrades also worked.
